File: admin/src/utils/productAvailability.ts
import type { BankHoliday, Product, ProductAvailability } from '../types';

export const AVAILABILITY_LABELS: Record<ProductAvailability, string> = {
  weekday: 'Weekday',
  weekend: 'Weekend',
  bank_holiday: 'Bank Holiday',
};

function dateKeyOf(date: Date): string {
  const y = date.getFullYear();
  const m = String(date.getMonth() + 1).padStart(2, '0');
  const d = String(date.getDate()).padStart(2, '0');
  return `${y}-${m}-${d}`;
}

// Bank holidays take priority over the plain weekday/weekend split -- a bank
// holiday that falls on a weekday (e.g. Christmas Day) is still a bank
// holiday for scheduling purposes, not an ordinary weekday.
export function dayTypeFor(date: Date, bankHolidays: BankHoliday[]): ProductAvailability {
  const key = dateKeyOf(date);
  // h.date is a UTC ISO string (e.g. a local midnight stored by the server
  // ends up as "...T23:00:00.000Z" for BST) -- comparing its raw slice(0,10)
  // would silently compare the wrong calendar day. Parse it back into a Date
  // first so the local getters undo that offset, same as everywhere else in
  // this codebase that round-trips a server date (e.g. BookingsPage's own
  // isSameDay).
  if (bankHolidays.some((h) => dateKeyOf(new Date(h.date)) === key)) return 'bank_holiday';
  const day = date.getDay();
  return day === 0 || day === 6 ? 'weekend' : 'weekday';
}

// Returns a warning message if the product's configured availability doesn't
// match the given date, or null if there's no restriction (or it matches).
export function availabilityMismatch(product: Product, date: Date, bankHolidays: BankHoliday[]): string | null {
  if (!product.availability) return null;
  const actual = dayTypeFor(date, bankHolidays);
  if (actual === product.availability) return null;
  return `${product.name} is set to only be used on a ${AVAILABILITY_LABELS[product.availability]}, but this date is a ${AVAILABILITY_LABELS[actual]}.`;
}

File: mobile/lib/utils/product_availability.dart
import 'package:flutter/material.dart';
import '../models/bank_holiday.dart';
import '../models/product.dart';

/// Human labels for a product's day-type restriction.
const availabilityLabels = {
  'weekday': 'Weekday',
  'weekend': 'Weekend',
  'bank_holiday': 'Bank Holiday',
};

String _dateKey(DateTime d) => '${d.year}-${d.month}-${d.day}';

/// The day-type of [date]: 'bank_holiday' takes priority over the plain
/// weekday/weekend split (a bank holiday on a weekday is still a bank holiday
/// for scheduling). Mirrors the admin app's dayTypeFor.
String dayTypeFor(DateTime date, List<BankHoliday> bankHolidays) {
  final key = _dateKey(DateTime(date.year, date.month, date.day));
  // A server date arrives as UTC ISO; round-trip through local so its calendar
  // day matches the local date, same as the admin util.
  final isHoliday = bankHolidays.any((h) => _dateKey(h.date.toLocal()) == key);
  if (isHoliday) return 'bank_holiday';
  final wd = date.weekday; // Mon=1..Sun=7
  return (wd == DateTime.saturday || wd == DateTime.sunday) ? 'weekend' : 'weekday';
}

/// A warning message if [product]'s restriction doesn't match [date]'s
/// day-type, or null if there's no restriction (or it matches).
String? availabilityMismatch(Product product, DateTime date, List<BankHoliday> bankHolidays) {
  final restriction = product.availability;
  if (restriction == null || restriction.isEmpty) return null;
  final actual = dayTypeFor(date, bankHolidays);
  if (actual == restriction) return null;
  return '${product.name} is set to only be used on a '
      '${availabilityLabels[restriction] ?? restriction}, but this date is a '
      '${availabilityLabels[actual] ?? actual}.';
}

/// Returns true if it's OK to use [product] on [date]: either there's no
/// day-type mismatch, or the user confirmed "Use anyway" on the warning.
/// Mirrors the admin "Check product type" modal.
Future<bool> confirmProductAvailability(
  BuildContext context,
  Product product,
  DateTime date,
  List<BankHoliday> bankHolidays,
) async {
  final message = availabilityMismatch(product, date, bankHolidays);
  if (message == null) return true;
  final ok = await showDialog<bool>(
    context: context,
    builder: (_) => AlertDialog(
      title: const Text('Check product type'),
      content: Text(message),
      actions: [
        TextButton(onPressed: () => Navigator.of(context).pop(false), child: const Text('Cancel')),
        FilledButton(onPressed: () => Navigator.of(context).pop(true), child: const Text('Use anyway')),
      ],
    ),
  );
  return ok ?? false;
}
